### ISSUE TITLE
Random 404 errors

### ISSUE THREAD

**Issue by shirhatti** [aspnet/HttpClientFactory#120]
_From @sepehr1014 on May 19, 2018 10:16_

On 2.1 RC1, we're getting occasional 404 errors from our API controllers (about 5 in 100). I was wondering if there's a known issue in this release and has a workaround. If not, I would appreciate tips on how to debug this issue and get to the root cause of these 404 errors.

_Copied from original issue: aspnet/KestrelHttpServer#2591_

**Comment by shirhatti** [comment]
_From @Tratcher on May 19, 2018 17:31_

No. I suggest you start by enabling logging. Fiddler may also be useful.

**Comment by shirhatti** [comment]
_From @sepehr1014 on May 19, 2018 18:47_

Thanks but we already log everything. The request does not reach the
action. We also have no filters in the pipeline. Which type of logging do
you mean?

**Comment by shirhatti** [comment]
_From @davidfowl on May 19, 2018 19:0_

Turning up the logging verbosity will tell you why the mvc action isn’t getting chosen. Most of our components have their own logging.

**Comment by shirhatti** [comment]
_From @sepehr1014 on May 19, 2018 19:2_

I just checked the full logs and spotted something odd. `Request.Method` is empty! How can that be? Shall I open another issue or track it in the current one?

**Comment by shirhatti** [comment]
_From @davidfowl on May 19, 2018 19:14_

This one is fine if that ends up being the root cause. Do you have the actual log? Can you paste it here? 

Also what is your environment setup?

**Comment by shirhatti** [comment]
_From @sepehr1014 on May 19, 2018 19:21_

We have a custom exception logger that stores some request properties
alongside some other useful data. It would be great if you could tell me
which properties to log ...
We're using 2.1 RC1 with IIS on Windows Server 2016. The odd thing is that
this started to happen about two days ago and the only recent fundamental
code change was the use of injected HttpClientFactory instead of a static
HttpClient. Can it have anything to do with this issue?

**Comment by shirhatti** [comment]
_From @sepehr1014 on May 22, 2018 10:48_

The culprit was definitely `HttpClientFactory`. As soon we went back to using a static HttpClient, we don't see this issue anymore.

**Comment by shirhatti** [comment]
_From @davidfowl on May 22, 2018 22:24_

/cc @rynowak @glennc

**Comment by rynowak** [comment]
@sepehr1014  - can you provide more information about this? What does the code look like that's calling `HttpClient`?

**Comment by sepehr1014** [comment]
@rynowak Sorry for my late response. It's injected into controllers and used as written in the official documentations. We have 2 out of ordinary usages though.

1. We have a messaging controller that has only one action called Connect which accepts a web socket request and is alive as long as the web socket is open. So we keep the injected `HttpClientFactory` alive for some time but we don't keep `HttpClient`s open. The factory is used like this:

```
using(var client = httpClientFactory.CreateClient())
    await client.PostAsync(...);
```

2. Sometimes we use the injected factory in a newly created long running thread:
```

Task.Factory.StartNew(async () =>
{
   using (var client = httpClientFactory.CreateClient())
             await client.PostAsJsonAsync(...);
}, TaskCreationOptions.LongRunning);
```

**Comment by DavidZidar** [comment]
I am also seeing random 404s in my application since I updated to ASP.NET Core 2.1 RC and they're still there in final. I have also been using HttpClientFactory so I tried removing it when I saw this thread but it did not help in my case.

The thing is, I'm not just seeing random 404s on ApiControllers but also static files. They happen every few hundred requests or so.

Unfortunately I have not been able to reproduce my issue in any other environment except my production environment which has actual traffic so it's very hard to debug.

Microsoft.AspNetCore.Builder.RouterMiddleware is logging "Request did not match any routes." even though clearly they should match.

Not sure if my issue is related, but it does sound similar.

**Comment by DavidZidar** [comment]
I finally solved my random 404s after many hours of debugging. ASP.NET Core was not at fault.

I am logging some performance metrics in aggregate in a middleware using Rx and a custom log framework and my log framework is using IHttpContextAccessor to extend log messages with HTTP-specifics when needed.

What I failed to consider was that Rx was capturing the thread somehow so my background perf events was using old HttpContext-instances and this is what seems to have caused my random 404s because HttpContext is indeed not thread safe.

It was a simple fix once I knew what the problem was. What I don't know why this wasn't a problem before ASP.NET Core 2.1.

**Comment by glennc** [comment]
@sepehr1014 So you are making an HTTP request from the code you pasted to one of your own APIs and getting a 404 sometimes even though everything is up? Are you doing any customization of the outgoing pipeline when using the factory or just adding it and using it without any extra stuff? This would be code in your configureServices method where you called AddHttpClient.

**Comment by sepehr1014** [comment]
@glennc No, we were actually using HttpClientFactory in our server code to send requests to Google Firebase servers (for FCM). When we had that in place, some of requests to our own APIs returned 404 error. So 404 had nothing to do with the requests sent from HttpClient but was somehow affected by the use of HttpClientFactory. We checked twice. When leveraging HttpClientFactory in our code, we saw a percentage of our API requests return 404.
